torch/csrc/dynamo/eval_frame.c

```diff
@@ -438,10 +484,14 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
     // If the hit cache entry is not the head of the linked list,
     // move it to the head
     if (prev != NULL) {
-        CacheEntry* extra = get_cache_entry(frame->f_code);
+        ExtraState* extra = get_extra_state(frame->f_code);
+        CacheEntry* old_cache_entry = extract_cache_entry(extra);
         prev->next = e->next;
-        e->next = extra;
-        set_cache_entry(frame->f_code, e);
+        e->next = old_cache_entry;
+
+        // Update the extra state to reflect the updated cache
+        extra->cache_entry = e;
```

Comment:
What if extra is NULL?

Reply:
It can't be NULL here. CacheEntry* e is not NULL, otherwise we would have returned at the beginning of this function. Since, CacheEntry is extracted from the extra, it can't be NULL here.